### PR TITLE
UNR-1993 Worker Metrics Delegate

### PR DESCRIPTION
**Pull request by joshuahuburn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Added a simple multicast delegate that users can bind to in `SpatialMetrics` when a Worker Metrics Op is received. 

#### Release note
Feature: Added a delegate to `SpatialMetrics` which triggers when worker metrics have been received.

#### Tests
Printed the result of the map passed to the delegate and it works.

#### Primary reviewers
@m-samiec 	@mattyoung-improbable

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialDispatcher.cpp [hunk 1]
```diff
@@ -117,6 +118,27 @@ void USpatialDispatcher::ProcessOps(Worker_OpList* OpList)
...
+void USpatialDispatcher::HandleWorkerMetrics(Worker_Op* Op)
+{
+	if (NetDriver->SpatialMetrics->WorkerMetricsRecieved.IsBound())
+	{
+		int32 NumMetrics = Op->metrics.metrics.gauge_metric_count;
+
+		if (NumMetrics > 0)
+		{
+			// Construct a map to store all the metrics and pass it to the users delegate
+			TMap<FString, double> WorkerMetrics;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialDispatcher.cpp]: Reserve the size of NumMetrics up front.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialDispatcher.cpp [hunk 2]
```diff
@@ -117,6 +118,27 @@ void USpatialDispatcher::ProcessOps(Worker_OpList* OpList)
 	Receiver->FlushRetryRPCs();
 }
 
+void USpatialDispatcher::HandleWorkerMetrics(Worker_Op* Op)
+{
+	if (NetDriver->SpatialMetrics->WorkerMetricsRecieved.IsBound())
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialDispatcher.cpp]: Probably worth just putting this logic inside the SpatialMetrics object.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialDispatcher.cpp [hunk 3]
```diff
@@ -102,8 +103,8 @@ void USpatialDispatcher::ProcessOps(Worker_OpList* OpList)
 			UE_LOG(LogSpatialView, Log, TEXT("SpatialOS Worker Log: %s"), UTF8_TO_TCHAR(Op->log_message.message));
 			break;
 		case WORKER_OP_TYPE_METRICS:
+			NetDriver->SpatialMetrics->HandleWorkerMetrics(Op);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialDispatcher.cpp]: Worth surrounding this in SHIPPING tags

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 4]
```diff
@@ -289,3 +289,25 @@ void USpatialMetrics::TrackSentRPC(UFunction* Function, ESchemaComponentType RPC
 	Stat.Calls++;
 	Stat.TotalPayload += PayloadSize;
 }
+
+void USpatialMetrics::HandleWorkerMetrics(Worker_Op* Op)
+{
+	if (WorkerMetricsRecieved.IsBound())
```
Comment (rokstrnisa) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp]: Typo

**Comment by m-samiec** [comment]
> Nice. Did we add Unreal GDK docs for this and informed Midwinter of the new feature?

We did - they reviewed the PR before it went in.

**Pull request merged**